nav2_costmap_2d/plugins/costmap_filters/keepout_filter.cpp

```diff
@@ -0,0 +1,256 @@
+      BASE_DEFAULT, MULTIPLIER_DEFAULT);
+  }
+
+  mask_topic_ = msg->map_mask_topic;
+
+  // Setting new map mask subscriber
+  RCLCPP_INFO(
+    node->get_logger(),
+    "Subscribing to \"%s\" topic for map mask...",
+    msg->map_mask_topic.c_str());
+  mask_sub_ = node->create_subscription<nav_msgs::msg::OccupancyGrid>(
+    mask_topic_, rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
+    std::bind(&KeepoutFilter::maskCallback, this, std::placeholders::_1));
+}
+
+void KeepoutFilter::maskCallback(
+  const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (mask_costmap_) {
+    RCLCPP_WARN(
+      node->get_logger(),
+      "New map mask arrived from %s topic. Updating old map mask.",
+      mask_topic_.c_str());
+    mask_costmap_.reset();
+  }
+
+  // Making a new mask_costmap_
+  mask_costmap_ = std::make_unique<Costmap2D>(*msg);
+}
+
+void KeepoutFilter::process(
+  nav2_costmap_2d::Costmap2D & master_grid,
+  int min_i, int min_j, int max_i, int max_j,
+  const geometry_msgs::msg::Pose2D & /*pose*/)
+{
+  std::lock_guard<CostmapFilter::mutex_t> guard(*getMutex());
+
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  if (!mask_costmap_) {
+    // Show warning message every 2 seconds to not litter an output
+    RCLCPP_WARN_THROTTLE(
+      node->get_logger(), *(node->get_clock()), 2000,
+      "Map mask was not received");
+    return;
+  }
+
+  double wx, wy;  // world coordinates
+
+  // Optimization: iterate only in overlapped
+  // (min_i, min_j)..(max_i, max_j) & mask_costmap_ area.
+  //
+  //           mask_costmap_
+  //       *----------------------------*
+  //       |                            |
+  //       |                            |
+  //       |      (2)                   |
+  // *-----+-------*                    |
+  // |     |///////|<- overlapped area  |
+  // |     |///////|   to iterate in    |
+  // |     *-------+--------------------*
+  // |    (1)      |
+  // |             |
+  // *-------------*
+  //  master_grid (min_i, min_j)..(max_i, max_j) window
+  //
+  // ToDo: after costmap rotation will be added, this should be re-worked.
+
+  // Calculating bounds corresponding to bottom-left overlapping (1) corner
+  int mg_min_x, mg_min_y;  // masger_grid indexes of bottom-left (1) corner
+  // mask_costmap_ -> master_grid intexes conversion
+  const double half_cell_size = 0.5 * mask_costmap_->getResolution();
+  wx = mask_costmap_->getOriginX() + half_cell_size;
+  wy = mask_costmap_->getOriginY() + half_cell_size;
+  master_grid.worldToMapNoBounds(wx, wy, mg_min_x, mg_min_y);
+  // Calculation of (1) corner bounds
+  if (mg_min_x >= max_i || mg_min_y >= max_j) {
```

Comment:
What does the origin have to do with the location of the costmap size overlap? That origin could be anywhere arbitrarily. For a global (non-rolling) costmap, they're completely arbitrary and have no relationship with the current update window. For a rolling-costmap, it could be related, but not promised to be so.

Reply:
This is not an origin of `master_grid`. This is the origin of `map_costmap_` (map_costmap is a `Costmap2D(mask_map)`). `wx` and `wy` - are world-coordinates of mask_costmap's origin, in other words - left bottom corner of `map_mask` I've called (1) in the picture in the comment in a code above. If that origin exceeds max_i/max_j window bounds, that means we have no overlapping with rolling window `(min_i, min_j, max_i, max_j)`. That is why this if-check is here.

Reply:
> are world-coordinates of mask_costmap's origin

But that could be arbitrarily anything, right? There's no promise that the origin of the mask is the bottom left corner, unless I'm missing something. The user could set the `origin` value in the yaml to nearly anything.

Reply:
Currently, there is no orientation for Costmap2D implemented yet. So, I've also supposed that all map masks have the the same as master grid orientation. The comment was added the optimization we are discussing that it is working only for non-rotational costmaps: https://github.com/ros-planning/navigation2/pull/1882/files/b36a5f3f74f1b799360e50e46715fa87b4310946#diff-f67a16397bb5979b7cbf4729ea53a451R169

Reply:
You are correct, sorry, my brain isn't doing well with this it seems. I'm used to thinking of the `origin` with respect to maps. In that respect, the `origin` is completely arbitrary, it could easily be `0,0` as much as `434235, -438230` and it's meaningless because that's the global starting point. When you have a keepout map that's _relative_ to some actual map, then the origin has to be meaningful to get the transformations correct and in that regard, origin is the `0,0` point in the lower left corner.